src/main/kotlin/pt/up/fe/ni/website/backend/model/dto/Dto.kt

```diff
@@ -0,0 +1,77 @@
+package pt.up.fe.ni.website.backend.model.dto
+
+import com.fasterxml.jackson.annotation.JsonIgnore
+import com.fasterxml.jackson.databind.ObjectMapper
+import javax.persistence.Entity
+import javax.validation.ConstraintViolationException
+import javax.validation.Validation
+import kotlin.reflect.KClass
+import kotlin.reflect.full.hasAnnotation
+import kotlin.reflect.jvm.jvmErasure
+
+open class Dto<T : Any> {
+
+    @JsonIgnore
+    private val entityClass: KClass<T>
+
+    constructor() : this(null)
+
+    constructor(conversionClass: KClass<T>?) {
+        this.entityClass = DtoReflectionUtils.getTypeConversionClassWithCache(this::class, conversionClass)
+    }
+
+    companion object {
+        private val objectMapper = ObjectMapper()
+        private val validatorFactory = Validation.buildDefaultValidatorFactory()
+    }
```

Comment:
I'm curious about `fields vs companion object`. Are there situations when one is preferred over the other? I'm asking for my own code too

Reply:
A companion object is the equivalent of a static class, that is a class that only exists once for all instances of Dto (you can think of it as putting static fields on Dto). That means that there will only be one validator factory, one object mapper and one conversion type cache across all Dto. In the case of ObjectMapper and Validator Factory, to reduce the number of those we have in our code, I've put them in the companion object. In the case of the type conversion cache, it has to be in the companion object since that information needs to be shared across instances of Dto.